src/sandstorm/grain.capnp

```diff
@@ -126,6 +126,99 @@ interface SandstormApi(AppObjectId) {
   # This method is here on `SandstormApi` rather than on `Identity` in order to ease a possible
   # future transition to a model where identity IDs are not globally stable and each grain has a
   # separate (possibly incompatible) map from identity ID to account ID.
+
+  schedule @11 ScheduledJob;
```

Comment:
Let's say the app calls `schedule()` and it throws a DISCONNECTED exception (or, really, any exception). The app doesn't really know for sure if the exception occurred before or after the scheduled job was persisted. What should it do? If it calls `schedule()` again, then it seems like it might end up scheduling two copies of the job, which seems problematic.

This is why we end up stuck with two-phase commit. :/

Alternatively, the other option is to make `schedule()` idempotent, so that retrying after error is always OK. That could mean that any time you schedule a job, any previously-scheduled job is removed. Or, if we feel that's too limiting and that a grain really needs to be able to schedule more than one job, we could make the grain assign names to jobs. Scheduling a new job with the same name as an existing job deletes the old job. We could also provide a way to list known jobs, which could allow an app to examine and clean up its own state.

Reply:
I think I prefer the idea of having the app provide an id for the job to avoid duplication vs. two-phase commit. Adding a method to list existing jobs seems like a good possible extension, though that could be a separate pr.

Reply:
It's kind of sad that this is sort of abandoning capability-based design by introducing a namespace, but I guess it is a lot easier to understand in this case.

I wonder if we should take it an extra step and not make the app provide a persistent callback, but instead have a method on `MainView` to invoke a scheduled job, passing the job ID...

Reply:
That *would* simplify the API substantially.

Conceptually its a bit sad, but it's not really operationally any different I think. With the current design you have:

1. App calls schedule(), passing a capability.
2. Sandstorm immediately asks the app for a token (bits) to represent that capability.
3. Sandstorm stores those bits.
4. Later, Sandstorm asks the app to turn those bits back into a capability, and calls run on it

Vs:

1. App calls schedule, passing a token (bits) representing the job.
2. Sandstorm stores those bits
3. Later, Sandstorm asks the app to interpret those bits and run a job.

I don't think the indirection of having a liveref really gains you anything here; the two are basically isomorphic, just with more steps in the current design.

Reply:
Good point that we don't actually need a new method on `MainView`... the app can just pass the appropriate SturdyRef at job creation.

I guess the fundamental problem with the "practical" approach here is separation of concerns. One can imagine a future where Sandstorm apps are commonly built on top of frameworks where deals of things like SturdyRefs are managed at the framework level while the application business logic code only concerns itself with LiveRefs. Such a framework will potentially need to special-case the scheduling API, if the scheduling API uses SturdyRefs explicitly. It could be kind of ugly.

On the other hand, with the purist approach, in the framework-happy future, the added tedium of implementing a persistent LifeRef will be handled by the framework so won't be a problem for apps.

But I dunno, I can definitely see the argument that we should design for the clear near term rather than a future than may never come. Your call I guess.